src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -5073,6 +5073,110 @@ .maxstack  1
+    public delegate*<string?, object?, C?> F5;
+    public delegate*<delegate*<string, int*>, delegate*<string?>, delegate*<void*, string>> F6;
+}";
+
+            var comp = CreateCompilationWithFunctionPointers(source, options: WithNonNullTypesTrue(TestOptions.UnsafeReleaseDll));
+            comp.VerifyDiagnostics();
+
+            verifySymbolNullabilities(comp.GetTypeByMetadataName("C")!);
+
+            CompileAndVerify(comp, symbolValidator: symbolValidator);
+
+            static void symbolValidator(ModuleSymbol module)
+            {
+                var c = module.GlobalNamespace.GetTypeMember("C");
+                Assert.Equal("System.Runtime.CompilerServices.NullableAttribute({0, 1, 1, 1})", getAttribute("F1"));
+                Assert.Equal("System.Runtime.CompilerServices.NullableAttribute({0, 1, 2, 1})", getAttribute("F2"));
+                Assert.Equal("System.Runtime.CompilerServices.NullableAttribute({0, 1, 1, 2})", getAttribute("F3"));
+                Assert.Equal("System.Runtime.CompilerServices.NullableAttribute({0, 2, 1, 1})", getAttribute("F4"));
+                Assert.Equal("System.Runtime.CompilerServices.NullableAttribute({0, 2, 2, 2})", getAttribute("F5"));
+                Assert.Equal("System.Runtime.CompilerServices.NullableAttribute({0, 0, 1, 0, 0, 0, 1, 0, 2})", getAttribute("F6"));
+
+                verifySymbolNullabilities(c);
+
+                string getAttribute(string fieldName) => c.GetField(fieldName).GetAttributes().Single().ToString()!;
+            }
+
+            static void verifySymbolNullabilities(NamedTypeSymbol c)
+            {
+                assertExpected("System.String!", "System.Object!", "C!", "F1");
+                assertExpected("System.String?", "System.Object!", "C!", "F2");
+                assertExpected("System.String!", "System.Object?", "C!", "F3");
+                assertExpected("System.String!", "System.Object!", "C?", "F4");
+                assertExpected("System.String?", "System.Object?", "C?", "F5");
+                assertExpected("System.String?", "System.Object?", "C?", "F5");
+                assertExpected("delegate*<System.String!, System.Int32*>", "delegate*<System.String?>", "delegate*<System.Void*, System.String!>", "F6");
+
+                void assertExpected(string param1Type, string param2Type, string returnType, string fieldName)
+                {
+                    var field = (FunctionPointerTypeSymbol)c.GetField(fieldName).Type;
+
+                    var paramTypes = field.Signature.ParameterTypesWithAnnotations;
+                    Assert.Equal(2, paramTypes.Length);
+                    Assert.Equal(param1Type, paramTypes[0].ToTestDisplayString(includeNonNullable: true));
+                    Assert.Equal(param2Type, paramTypes[1].ToTestDisplayString(includeNonNullable: true));
+
+                    Assert.Equal(returnType, field.Signature.ReturnTypeWithAnnotations.ToTestDisplayString(includeNonNullable: true));
```

Comment:
Why not use `Assert.Equal("...", field.ToTestDisplayString(includeNonNullable: true))` rather than asserting each part separately?

Reply:
Because these were the first tests I wrote for decoding, and I hadn't worked out the good way to test them yet. I can update it.

---
In reply to: [420565834](https://github.com/dotnet/roslyn/pull/43961#discussion_r420565834) [](ancestors = 420565834)